### PR TITLE
Jcn 505 modify sqs package to be able to send messages above the limit

### PR DESCRIPTION
**Pull request by merq-rodriguez**
## Descripción

Con los cambios implementados en esta épica, buscamos permitir el intercambio de payloads entre recursos sin preocuparnos por las limitaciones impuestas por AWS (principalmente el límite de 256 KB).

Para lograrlo, almacenaremos todos los payloads recibidos en un bucket S3 de un nuevo microservicio que solo contiene los buckets para estos procesos. La idea es subir los archivos, organizándolos en rutas que indiquen el tipo y nombre del recurso que generó el input. Luego, el recurso encargado de procesar ese payload recibirá la URL para descargar y procesar el contenido.

## Summary by CodeRabbit

- **New Features**
  - Introduced a robust SQS event publisher supporting single and batch event publishing, with automatic handling of large messages via S3 offloading.
  - Added concurrency control for batch publishing and detailed response reporting.
  - Exposed new helpers for random value generation, property picking, AWS parameter store access, S3 uploads, and required IAM permissions.
  - Published a set of predefined error codes for easier troubleshooting.

- **Documentation**
  - Significantly expanded and clarified the README with detailed usage instructions, compatibility notes, and code examples.
  - Fixed the coverage badge branch in documentation.

- **Tests**
  - Added comprehensive tests for all major helpers and the SQS event publisher, including edge cases and error handling.
  - Removed obsolete empty test file and replaced with a fully implemented test suite for the SQS event publisher.

- **Chores**
  - Updated and added dependencies to support new AWS SDK integrations and testing utilities.
  - Added TypeScript declaration build script and updated `.gitignore` to exclude generated types.

### CODE REVIEW

File: lib/helpers/assume-role.js [hunk 1]
```diff
@@ -0,0 +1,64 @@
+'use strict';
+
+const { STSClient, AssumeRoleCommand } = require('@aws-sdk/client-sts');
+const SqsEmitterError = require('../sqs-emitter-error');
+
+module.exports = class AssumeRole {
+
+	/**
+    * @private
+    * @static
```
Comment (Alejandro-Gonzalez) [inline lib/helpers/assume-role.js]: @merq-rodriguez Fijate que tenes espacios en vez de tabs aca y en el siguiente method

File: lib/helpers/async-with-concurrency.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+module.exports = class AsyncWithConcurrency {
+
+	constructor(workerFn, maxConcurrency) {
+		this.workerFn = workerFn;
+		this.maxConcurrency = maxConcurrency;
+		this.activeTasks = 0;
+	}
+
+	async run(queue) {
```
Comment (Alejandro-Gonzalez) [inline lib/helpers/async-with-concurrency.js:13]: @merq-rodriguez con este metodo tengo dudas, esta medio raro como lo hiciste, con el this.activeTasks y con el promise.race, por que supone que la primera vez a while que entra al while y tenes 2 tasks esta mete 2 promises y luego entra otra vez la primera vez y en la segunda otra vez se hace otra por que this.activeTask es 1 todavia?
No podriamos no depender de activeTask y usar un promise.allSettled?

File: lib/helpers/async-with-concurrency.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+		// Make this copy to avoid modifying the original queue
+		const queueCopy = [...queue];
+
+		const results = [];
+		const promises = new Map();
+		let itemIndex = 0;
+
+		while(queueCopy.length || this.activeTasks) {
```
Comment (Alejandro-Gonzalez) [inline lib/helpers/async-with-concurrency.js:22]: @merq-rodriguez algo asi capas o el promise race tiene un porque?
```js
while(queueCopy.length) {

	const chunkConcurrency = Math.min(this.maxConcurrency, queueCopy.length);

	const tasks = queueCopy.splice(0, chunkConcurrency);

	tasks.forEach(task => {
		promises.set(itemIndex, this.processTask(task, itemIndex));
		itemIndex++;
	});

	const promiseResults = await Promise.allSettled(promises.values());

	for(const { value, status, reason } of promiseResults) {

		if(status === 'rejected')
			throw reason;

		results[value.index] = value.value;
		promises.delete(value.index);
	}
```

File: lib/helpers/s3-uploader.js [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+				Body: body
+			}));
+
+		} catch(error) {
+			logger.error(`Error uploading to bucket ${bucket.bucketName} in region ${bucket.region}: ${error.message}`);
+		}
+	}
+
+	/**
+ * Handle uploads content to a specified S3 content path, attempting first on the default bucket, and then retrying on a provisional bucket if the first attempt fails.
```
Comment (Alejandro-Gonzalez) [inline lib/helpers/s3-uploader.js]: @merq-rodriguez aca metele un tab aca en el jsdoc para que quede prolijo

File: lib/sqs-emitter.js [hunk 5]
```diff
@@ -1,11 +1,331 @@
...
+	async publishEvents(sqsUrl, events) {
+
+		const parsedEvents = this.parseEvents(events, sqsUrl);
+
+		const asyncWithConcurrency = new AsyncWithConcurrency(async batch => {
+
+			const promises = [];
+			const formattedSqsBatch = [];
+
+			for(const { limitExceeded, extraProperties: { payloadFixedProperties, s3ContentPath }, ...parsedEvent } of batch) {
```
Comment (Alejandro-Gonzalez) [inline lib/sqs-emitter.js]: @merq-rodriguez tanto aca como en el publishEvent hacemos los mismo para formatear los eventos que exceden el limiete y para subirlos a s3, podemos meter unos ajustes para meter una funcion para retulizar codigo.
Algo asi donde te o formatea y te devuelve la promise para que la utilices en tu logica 

```js
async formatAndUploadEventWithS3Content(parsedEvent) {

	const { limitExceeded, extraProperties: { payloadFixedProperties, s3ContentPath }, ...parsedEventBase } = parsedEvent;

	const bucketList = await ParameterStore.getParameterValue();

	const sqsFixedContent = this.formatBodyWithS3ContentPath(parsedEvent, s3ContentPath, payloadFixedProperties);

	return {
		formattedEvent: { ...parsedEventBase, MessageBody: JSON.stringify(sqsFixedContent) },
		s3UploadRequest: S3Uploader.uploadS3ContentPath(bucketList, s3ContentPath, parsedEvent.MessageBody)
	};
}
```

File: lib/sqs-emitter.js [hunk 6]
```diff
@@ -1,11 +1,331 @@
...
+			QueueUrl: sqsUrl
+		}));
+
+		return {
+			messageId: sqsResponse.MessageId,
+			...sqsResponse.SequenceNumber && { sequenceNumber: sqsResponse.SequenceNumber }
+		};
+	}
+
+	async publishEvents(sqsUrl, events) {
```
Comment (Alejandro-Gonzalez) [inline lib/sqs-emitter.js:76]: @merq-rodriguez Aca te prodria quedar algo asi

```js
for(const { limitExceeded, extraProperties, ...parsedEvent } of batch) {

	if(!limitExceeded) {
		formattedSqsBatch.push(parsedEvent);
		continue;
	}

	const { formattedEvent, s3UploadRequest } = await this.formatAndUploadEventWithS3Content({ ...parsedEvent, extraProperties });

	formattedSqsBatch.push(formattedEvent);

	promises.push(s3UploadRequest);
}

if(promises.length)
	await Promise.all(promises);
```

File: lib/sqs-emitter.js [hunk 7]
```diff
@@ -1,11 +1,331 @@
...
+		return randomValue(13);
+	}
+
+	async publishEvent(sqsUrl, event) {
+
+		const sqsName = this.getSqsNameFromUrl(sqsUrl);
+		let parsedEvent = this.formatSQSEvent(event, sqsName);
+		let parsedEventSize = JSON.stringify(parsedEvent).length;
+
+		const {
```
Comment (Alejandro-Gonzalez) [inline lib/sqs-emitter.js]: @merq-rodriguez aca te podria quedar asi

```js
const { extraProperties, ...parsedEventBase } = parsedEvent;

let formattedSqsMessage = parsedEventBase;

[parsedEvent, parsedEventSize] = this.handleEventSizeLimit(parsedEvent, parsedEventSize);

if(parsedEvent.limitExceeded) {

	const { formattedEvent, s3UploadRequest } = await this.formatAndUploadEventWithS3Content(parsedEvent);

	formattedSqsMessage = formattedEvent;

	await s3UploadRequest;
}
```

File: lib/sqs-emitter.js [hunk 8]
```diff
@@ -1,11 +1,331 @@
...
+const logger = require('lllog')();
+const { SQSClient, SendMessageBatchCommand, SendMessageCommand } = require('@aws-sdk/client-sqs');
+
+const AsyncWithConcurrency = require('./helpers/async-with-concurrency');
+const ParameterStore = require('./helpers/parameter-store');
+const S3Uploader = require('./helpers/s3-uploader');
+const { randomValue } = require('./helpers/id-helper');
+const { pickProperties } = require('./helpers/pick-properties');
+const SqsEmitterError = require('./sqs-emitter-error');
 // const SqsEmitterError = require('./sqs-emitter-error');
```
Comment (Alejandro-Gonzalez) [inline lib/sqs-emitter.js]: @merq-rodriguez te quedo este comentario aca!

**Review by Alejandro-Gonzalez — CHANGES_REQUESTED** [review]

File: lib/helpers/async-with-concurrency.js (hunk 3 above)
Comment (merq-rodriguez) [reply lib/helpers/async-with-concurrency.js:22]: Buenas! Como va! @Alejandro-Gonzalez

Te comento, esta clase es parte de la implementación que hizo @jormaechea en el package de sns, el `Promise.race`, este método espera a que la primera promesa de un conjunto se resuelva (o rechace) y devuelve su resultado. Esto permite que, en lugar de esperar a que todas las tareas terminen (como hace Promise.all), el servicio pueda ir procesando tareas a medida que se completan. Si algunas tareas son más rápidas que otras, estas se resuelven primero, lo que puede mejorar la eficiencia en escenarios donde los tiempos de ejecución varían. Por ejemplo, subir un pedido de 5 MB a S3 toma más tiempo que uno de 200 KB.

**AsyncWithConcurrency** con **Promise.race**: La clase limita la concurrencia con `maxConcurrency`. Si tienes 5 tareas y `maxConcurrency = 2`, se ejecutan 2 tareas a la vez, y cuando una termina, se lanza la siguiente. El tiempo total depende del número de tareas, el límite de concurrencia que seria `25` (hasta 25 tareas en paralelo) y el tiempo por tarea. Se enfatiza en el control al limitar la concurrencia para no saturar el servicio, termina 1 tarea, se indexa en los resultados, se elimina de las promesas y si quedan mas en la cola, otra tomara ese lugar.

Si llegaste hasta aqui, un saludo jaja y si vez algo que no te termine de encajar comentalo y lo miramos!!!!
Capaz podamos mejorar los tiempos y el control de las tareas! (Y)

Comment (Alejandro-Gonzalez) [reply lib/helpers/async-with-concurrency.js:22]: Sisi entiendo, ahora mas, para lo que vamos a hacer es suficiente supongo, pero tambien veo que cada task puede tener n promises, por lo que cada task no siempre es una promise y entonces podrias tener mas promises que maximas concurrencias, pero no se si es tan critico. Capas ese 25 puede ser menor nose.  Lo dejo como observacion nomas.

**Review by Alejandro-Gonzalez — APPROVED** [review]

**Review by gastonpereyra — APPROVED** [review]

**Review by jormaechea — CHANGES_REQUESTED** [review]
Ver los comentarios de https://github.com/janis-commerce/sns/pull/2

File: lib/helpers/s3-uploader.js [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+ * @param {string} contentS3Path - The path in the S3 bucket where the content will be uploaded.
+ * @param {string} body - The payload to upload to s3 bucket
+ * @returns {Promise<void>} - Resolves if the content is successfully uploaded to either the default or provisional bucket.
+ * Rejects with an error if both attempts fail.
+ *
+ * @throws {Error} - If the upload fails for both the default and provisional buckets, an error is thrown.
+ */
+	static async uploadContentS3Path(buckets, contentS3Path, body) {
+
+		const [defaultBucket, provisionalBucket] = buckets;
```
Comment (manuelvilche) [inline lib/helpers/s3-uploader.js]: @merq-rodriguez Aca estaria bueno recorrer ese array de buckets siempre tomando como primera opcion el default

**Review by manuelvilche — CHANGES_REQUESTED** [review]

**Comment by jormaechea** [comment]
@coderabbitai review

File: README.md [hunk 10]
```diff
@@ -1,23 +1,118 @@
...
 npm install @janiscommerce/sqs-emitter
 ```
 
-## API
-
+### Install peer dependencies
+```sh
+# Install as devDependency if you run your code in AWS Lambda, which already includes the SDK
+npm install --dev @aws-sdk/client-sqs@3
+```
```
Comment (jormaechea) [inline README.md]: Esto no hace falta, porque esta agregado como dependencia...

File: README.md [hunk 11]
```diff
@@ -1,23 +1,118 @@
...
 
 ## Usage
 ```js
 const SqsEmitter = require('@janiscommerce/sqs-emitter');
 
 ```
 
-## Examples
+> Why? This is to avoid installing the SDK in production and freezing the SDK version in this package
```
Comment (jormaechea) [inline README.md]: Esta linea tampoco tiene sentido ya..

File: README.md [hunk 12]
```diff
@@ -1,23 +1,118 @@
...
+
+> This class is compatible with [@janiscommerce/api-session](https://npmjs.com/@janiscommerce/api-session). If it's instanciated using `getSessionInstance`, a message attribute `janis-client` with the session's `clientCode` will be automatically added to every event.
+
+> The event `content` will be JSON-stringified before sending
+
+> The event `attributes` can be either Strings or Arrays.  It's important to note that using other data types may cause issues or inconsistencies in the implemented filter policies. Ensure that the values provided for the attributes are always of the expected type to avoid errors in message processing.
+
+> The `payloadFixedProperties` event must be an array of strings containing the properties that must be mandatorily sent in the content. This is to improve error management, as these properties will allow us to identify which data failed and make a decision accordingly.
+
+**!important:** The session is required to obtain the `clientCode` and construct the `contentS3Path` for payloads that exceed the maximum SQS message size limit.
```
Comment (jormaechea) [inline README.md]: Quedó raro este `!important`...
image

File: README.md [hunk 13]
```diff
@@ -1,23 +1,118 @@
...
+
+#### Publish single event
+
+```js
+const { SqsEmitter } = require('@janiscommerce/sqs-emitter');
+
+const sqsEmitter = this.session.getSessionInstance(SqsEmitter);
+
+const result = await sqsEmitter.publishEvent('http://sqs-url', {
```
Comment (jormaechea) [inline README.md]: Pongamos una URL que tenga mas sentido (sin publicar un ID de cuenta de AWS real, pero que el formato sea el de una Queue)

File: README.md [hunk 14]
```diff
@@ -1,23 +1,118 @@
...
+#### Publish multiple events
+
+> This method will send multiple events in one SDK call. It will also separate in batches when the total size limit of 256KB payload size is exceeded. Batches will be sent with a smart concurrency protocol (optimizing calls with a maximum of 25 concurrent calls).
+
+```js
+const { SqsEmitter } = require('@janiscommerce/sqs-emitter');
+
+const sqsEmitter = this.session.getSessionInstance(SqsEmitter);
+
+const result = await sqsEmitter.publishEvents('http://sqs-url', [
```
Comment (jormaechea) [inline README.md]: Acá lo mismo de la URL..

File: lib/helpers/parameter-store.js [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+ * @property {string} region - The AWS region where the bucket is located.
+ * @property {boolean} default - Indicates if this bucket is the default (optional).
+ */
+
+class ParameterStore {
+
+	/**
+		* @private
+		* @static
+		*/
```
Comment (jormaechea) [inline lib/helpers/parameter-store.js]: Quedaron mal indentados todos los JSDocs.. Hay un tab donde deberia ser un espacio...

File: lib/helpers/pick-properties.js [hunk 16]
```diff
@@ -0,0 +1,13 @@
...
+
+	const result = {};
+
+	for(const key of keys) {
+		if(object[key])
+			result[key] = object[key];
+	}
+
+	return result;
+};
```
[the remark this answers is not in the document]
Comment (jormaechea) [reply lib/helpers/pick-properties.js]: Tiene sentido lo que comenta el 🐰 , pero creo que iria mas por simplemente `if(typeof object[key] !== 'undefined')` y fue...

File: lib/sqs-emitter.js [hunk 17]
```diff
@@ -1,11 +1,365 @@
...
+		const parsedEvents = this.parseEvents(events, sqsUrl);
+
+		const asyncWithConcurrency = new AsyncWithConcurrency(async batch => {
+
+			const formattedSqsPromises = [];
+
+			for(const { limitExceeded, extraProperties, ...parsedEvent } of batch) {
+
+				if(!limitExceeded) {
+					formattedSqsPromises.push(parsedEvent);
```
Comment (jormaechea) [inline lib/sqs-emitter.js]: `parsedEvent` es una Promise? No encontré donde es que es una promise...

File: lib/sqs-emitter.js [hunk 18]
```diff
@@ -1,11 +1,365 @@
...
+				DataType: 'String',
+				StringValue: this.session.clientCode
+			};
+		}
+
+		if(attributes) {
+			Object.entries(attributes).forEach(([key, value]) => {
+				parsedAttributes[key] = {
+					DataType: Array.isArray(value) ? 'String.Array' : 'String',
+					StringValue: Array.isArray(value) ? JSON.stringify(value) : value
```
Comment (jormaechea) [inline lib/sqs-emitter.js]: Agreguemos un caso mas, que nos paso en varios lugares.. Que si `value` es un boolean, mandarlo como `value.toString()`..

Sino en cada lugar que se usa esto hay que hacer `myAttribute: myBool.toString()` en lugar de simplemente `myAttribute: myBool`

File: lib/sqs-emitter.js [hunk 19]
```diff
@@ -1,11 +1,365 @@
...
+				};
+			});
+		}
+
+		return parsedAttributes;
+	}
+
+	parseExtraProperties(event, sqsName) {
+
+		if(!this.session?.clientCode)
```
Comment (jormaechea) [inline lib/sqs-emitter.js]: Esto me parece raro... No puedo tener un mensaje que exceda los 256MB y que no tenga un cliente asociado? 🤔

File: tests/helpers/pick-properties.js [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+
+	it('should return empty object when keys array is empty', () => {
+		const input = { a: 1, b: 2 };
+		const keys = [];
+		const expected = {};
+		const result = pickProperties(input, keys);
+		assert.deepStrictEqual(result, expected);
+	});
+
+	it('should skip falsy values', () => {
```
Comment (jormaechea) [inline tests/helpers/pick-properties.js]: `should skip falsy values` no estoy muy de acuerdo de que deba hacerlo 😅

File: tests/sqs-emitter.js [hunk 21]
```diff
@@ -0,0 +1,958 @@
...
+			assert.deepStrictEqual(s3Mock.commandCalls(PutObjectCommand).length, 0);
+			assert.deepStrictEqual(sqsMock.commandCalls(SendMessageBatchCommand).length, 0);
+			assert.deepEqual(s3Mock.commandCalls(PutObjectCommand).length, 0);
+
+			assertRamListResourceCommand();
+			assertSsmGetParameterCommand();
+
+		});
+
+		it('Should reject if payload upload fails in both S3 buckets', async () => {
```
Comment (jormaechea) [inline tests/sqs-emitter.js]: Tiene sentido que explote si no puede subir uno de los 100 mensajes que le paso? No deberia simplemente fallar ese mensaje?

**Review by jormaechea — CHANGES_REQUESTED** [review]

File: lib/helpers/parameter-store.js [hunk 22]
```diff
@@ -0,0 +1,103 @@
...
+		* @returns {Promise<string>} The ARN of the filtered resource that matches the parameter name.
+		* @throws {Error} If there is an error while listing RAM resources or no resources match the parameter.
+		*/
+	static async getParameterArnFromRAM() {
+
+		try {
+
+			const command = new ListResourcesCommand({ resourceOwner: 'OTHER-ACCOUNTS' });
+
+			const ramClient = new RAMClient({ region: 'us-east-1' });
```
Comment (jormaechea) [inline lib/helpers/parameter-store.js]: Esta raro que estemos hardcodeando la region (aca y en cualquier lado)... Creo que deberiamos dejarlo vacío y que tome la region actual..

File: lib/helpers/permissions.js [hunk 23]
```diff
@@ -0,0 +1,14 @@
+'use strict';
+
+const sqsPermissions = [
+	{
+		action: ['ssm:GetParameter'],
+		resource: 'arn:aws:ssm:us-east-1:*:parameter/shared/internal-storage'
```
[the remark this answers is not in the document]
Comment (jormaechea) [reply lib/helpers/permissions.js]: Lo reemplazaría por `resource: 'arn:aws:ssm:${aws:region}:*:parameter/shared/internal-storage'`

File: lib/sqs-emitter.js (hunk 19 above)
Comment (merq-rodriguez) [reply lib/sqs-emitter.js]: Agregue la validación por si se instancia la clase sin inyectar una session tipo `this.session.getSessionInstance(SqsEmitter)`, pero si lo vez muy redundante quitamos la validación :)

File: tests/helpers/pick-properties.js (hunk 20 above)
Comment (merq-rodriguez) [reply tests/helpers/pick-properties.js]: Que te parece 'Should skip undefined values' ?
Es acorde a la validación que se agregó, aunque aun estamos dejando pasar valores con **null** no se si debamos... @jormaechea  O.o

File: lib/helpers/parameter-store.js (hunk 15 above)
Comment (jormaechea) [reply lib/helpers/parameter-store.js]: Esto sigue mal...

File: lib/sqs-emitter.js (hunk 19 above)
Comment (jormaechea) [reply lib/sqs-emitter.js]: Pero que pasaría para un mensaje de una entidad Core, que no tiene cliente? No deberiamos poder manejarlo y guardar en un directorio especial para los mensajes sin cliente?

Comment (merq-rodriguez) [reply lib/sqs-emitter.js]: No habia visto tu respuesta @jormaechea, actualmente tenemos manejo para entidades con cliente algo como esto:

`sqsContent/${clientCode}/service-name/MySQSName/2025/03/06/${randomId}.json
`

Para entidades core o algo que no tenga un cliente que path podriamos usar, algo como esto?

`sqsContent/ custom | core | generic | root | storage | misc /service-name/MySQSName/2025/03/06/${randomId}.json
`

Seria cambiar el clientCode por algo diferente y un tanto generico, como lo vez, se te ocurre algo?

De momento voy a dejar `storage` para probar...

File: lib/helpers/parameter-store.js [hunk 24]
```diff
@@ -0,0 +1,101 @@
...
+	 *
+	 * @returns {Promise<Array<S3BucketConfig> | undefined>}
+	 * Returns a parsed JSON array of objects, where each object contains
+	 * information about S3 buckets, such as the `bucketName`, `roleArn`, `region`, and an optional `default` flag.
+	 * If the parameter retrieval fails, it logs an error and returns `undefined`.
+	 * @throws {Error} - If the `getParameterArnFromRAM` method fails, or the SSM command fails.
+	 */
+	static async getParameterValue() {
+
+		const parameterArn = await this.getParameterArnFromRAM();
```
[the remark this answers is not in the document]
Comment (jormaechea) [reply lib/helpers/parameter-store.js:47]: Do not report `lint/complexity/noThisInStatic` any more for this orgenization

File: lib/sqs-emitter.js [hunk 25]
```diff
@@ -1,11 +1,403 @@
...
+		return {
+			...eventIndex && { Id: `${eventIndex}` },
+			MessageBody: JSON.stringify(event.content),
+			MessageAttributes: parsedAttributes,
+			...event.subject && { Subject: event.subject },
+			...event.messageDeduplicationId && { MessageDeduplicationId: event.messageDeduplicationId },
+			...event.messageGroupId && { MessageGroupId: event.messageGroupId },
+			...event.messageStructure && { MessageStructure: event.messageStructure },
+			...parsedAttributes && { MessageAttributes: parsedAttributes },
+			...extraProperties && { extraProperties }
```
[the remark this answers is not in the document]
Comment (jormaechea) [reply lib/sqs-emitter.js]: @merq-rodriguez ⬆️

File: lib/sqs-emitter.js [hunk 26]
```diff
@@ -1,11 +1,399 @@
...
+			Object.entries(attributes).forEach(([key, value]) => {
+
+				let StringValue;
+
+				if(Array.isArray(value))
+					StringValue = JSON.stringify(value);
+				else if(typeof value === 'boolean')
+					StringValue = value.toString();
+				else
+					StringValue = value;
```
[the remark this answers is not in the document]
Comment (jormaechea) [reply lib/sqs-emitter.js]: @merq-rodriguez ⬆️ 

O lo que sugirió, o simplemente

```js
if(isArray)
  JSON.stringify(value);
else
  String(value);
```

Porque el boolean deja de ser un caso especial..

File: lib/sqs-emitter.js [hunk 27]
```diff
@@ -1,11 +1,378 @@
...
+
+				parsedAttributes[key] = {
+					DataType: Array.isArray(value) ? 'String.Array' : 'String',
+					StringValue
+				};
```
[the remark this answers is not in the document]
Comment (jormaechea) [reply lib/sqs-emitter.js]: @merq-rodriguez ⬆️

File: lib/sqs-emitter.js [hunk 28]
```diff
@@ -1,11 +1,365 @@
...
+		if(sqsName.endsWith('.fifo'))
+			return sqsName.substring(0, sqsName.length - 5);
+
+		return sqsName;
+	}
+
+	isValidSqsUrl(url) {
+		const pattern = /^https:\/\/sqs\.[a-zA-Z0-9-]+\.amazonaws\.com\/\d{12}\/[a-zA-Z0-9]+(\.fifo)?$/;
+		return pattern.test(url);
+	}
```
[the remark this answers is not in the document]
Comment (jormaechea) [reply lib/sqs-emitter.js]: @merq-rodriguez ⬆️ 

No me preocuparia tanto por el tema del largo, pero si que valide que el nombre de la queue pueda tener guiones..

Cambiaría solo `[a-zA-Z0-9]+(\.fifo)?` por `[a-zA-Z0-9_-]+(\.fifo)?` (agregar `_-` al nombre de la queue)

**Review by jormaechea — CHANGES_REQUESTED** [review]
@merq-rodriguez quedaron varias correcciones sin solucionar.. Y hay un test rompiendo tambien...

File: lib/sqs-emitter.js (hunk 27 above)
[the remark this answers is not in the document]
Comment (merq-rodriguez) [reply lib/sqs-emitter.js]: Yo estuve probando este caso ayer y el SDK no rompe, envia el mensaje porque el array esta convertido a string, como que no es estricto con los tipos, llega asi:

image
